core/src/main/java/com/amplifyframework/util/UserAgent.java

```diff
@@ -20,26 +20,83 @@
+import com.amplifyframework.core.AmplifyConfiguration;
 import com.amplifyframework.core.BuildConfig;
+import com.amplifyframework.logging.Logger;
+
+import java.util.Map;
 
 /**
  * A utility to construct a User-Agent header, to be sent with all network operations.
  */
 public final class UserAgent {
+    private static final Logger LOG = Amplify.Logging.forNamespace("amplify:core");
+    private static final int SIZE_LIMIT = 254; // VARCHAR(254)
+
     private static String instance = null;
 
     private UserAgent() {}
 
+    /**
+     * Configure User-Agent singleton with Amplify configuration instance.
+     * @param configuration An Amplify configuration instance.
+     * @throws AmplifyException If called twice or user-agent exceeds size limit.
+     */
+    public static synchronized void configure(@NonNull AmplifyConfiguration configuration)
+            throws AmplifyException {
+        // Block any sub-sequent configuration call.
+        if (instance != null) {
+            throw new AmplifyException(
+                    "User-Agent was already configured successfully.",
+                    "User-Agent is configured internally during Amplify configuration. " +
+                            "This method should not be called externally."
+            );
+        }
+
+        // Pre-pend the additional platforms before Android user-agent
+        final StringBuilder userAgent = new StringBuilder();
+        for (Map.Entry<Platform, String> platform : configuration.getAdditionalPlatforms().entrySet()) {
+            userAgent.append(String.format("%s/%s ",
+                    platform.getKey().getLibraryName(),
+                    platform.getValue()));
+        }
+        userAgent.append(androidUserAgent());
+
+        // The character limit for our User-Agent header is 254 characters.
+        // HTTP does not impose a maximum, but the AWS SDKs & Tools database
+        // that stores metrics records declares user-agent as a VARCHAR(254)
+        if (userAgent.length() > SIZE_LIMIT) {
+            throw new AmplifyException(
+                    "User-Agent exceeds the size limit of VARCHAR(254).",
+                    AmplifyException.REPORT_BUG_TO_AWS_SUGGESTION
+            );
+        }
+
+        instance = userAgent.toString();
+    }
+
     /**
      * Gets a String to use as the value of a User-Agent header.
      * @return A value for a User-Agent header.
+     * @throws RuntimeException If called before {@link #configure(AmplifyConfiguration)}.
      */
     @SuppressLint("SyntheticAccessor")
     @NonNull
     public static String string() {
         if (instance == null) {
-            instance = new UserAgent.Builder()
-                .libraryName("amplify-android")
+            LOG.debug("User-Agent is not yet configured. Returning default Android user-agent.");
+            return androidUserAgent();
+        }
+
+        return instance;
+    }
+
+    private static String androidUserAgent() {
```

Comment:
should this just be part of `string()`?

Reply:
its used twice in the code so i just made it a standalone method. renamed to `forAndroid()` though. (`UserAgent.forAndroid()` sounds better than `UserAgent.androidUserAgent()`)